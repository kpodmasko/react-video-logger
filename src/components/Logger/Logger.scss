.logger {
  position: relative;

  &__overlay {
    position: absolute;
    pointer-events: none;
    top: 0;
    left: 0;
    z-index: 10;
    overflow: hidden;
  }
}
