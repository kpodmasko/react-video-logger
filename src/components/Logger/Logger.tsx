import React, {
  FC,
  memo,
  ReactElement,
  useCallback,
  useEffect,
  useRef,
  ReactChild,
  useState,
} from "react";
import erd from "element-resize-detector";
import classNames from "classnames";

import { Classes, LogInfo, Size, Time } from "@declarations/types";
import { IBaseProps } from "@declarations/interfaces";

import "./Logger.scss";

interface ILoggerProps extends IBaseProps {
  logs: Array<LogInfo>;
  children: ReactChild | ReactElement | null; // expects only 1 child to calc logger width and height
  currentTime: Time;
}

type CanvasSizes = {
  width: Size;
  height: Size;
};

export const mainCssClass = "logger";

// TODO: add tests
const Logger: FC<ILoggerProps> = (props: ILoggerProps) => {
  const { className, logs, children, currentTime } = props;

  const [sizes, setSizes] = useState<CanvasSizes>({ width: 0, height: 0 });

  const loggerSourceWrapperRef = useRef<HTMLDivElement>(null);
  const loggerOverlayRef = useRef<HTMLCanvasElement>(null);

  const classes: Classes = classNames(mainCssClass, className);

  const handleResize = useCallback((element: HTMLElement) => {
    const { offsetWidth: width, offsetHeight: height } = element;
    setSizes({
      width,
      height,
    });
  }, []);

  useEffect(() => {
    const loggerChildInstance: Element =
      loggerSourceWrapperRef.current?.firstElementChild;

    if (!loggerChildInstance) {
      return;
    }

    const resizeListener = erd();
    resizeListener.listenTo(loggerChildInstance, handleResize);

    return () => {
      resizeListener.uninstall(loggerChildInstance);
    };
  }, [handleResize]);

  useEffect(() => {
    const canvas: HTMLCanvasElement = loggerOverlayRef.current;
    const context: CanvasRenderingContext2D = canvas.getContext("2d");
    context.clearRect(0, 0, canvas.width, canvas.height);

    logs?.forEach(({ zone }) => {
      const { top, left, width, height } = zone;

      context.beginPath();
      context.rect(left, top, width, height);
      context.lineWidth = 15;
      context.strokeStyle = "green";
      context.stroke();
    });
  }, [logs, currentTime]);

  return (
    <div className={classes}>
      <div
        className={`${mainCssClass}__source-wrapper`}
        ref={loggerSourceWrapperRef}
      >
        {children}
      </div>
      <canvas
        className={`${mainCssClass}__overlay`}
        {...sizes}
        ref={loggerOverlayRef}
      />
    </div>
  );
};

export default memo(Logger);
