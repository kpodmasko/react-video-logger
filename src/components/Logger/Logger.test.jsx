import React from "react";
import { shallow, mount } from "enzyme";

import Logger from "@components/Logger";
import { mainCssClass } from "@components/Logger/Logger";

describe("Logger", () => {
  it("renders without error", () => {
    shallow(<Logger />);
  });

  it("implements className prop", () => {
    const className = "test-class-name";
    const wrapper = shallow(<Logger className={className} />);

    expect(wrapper.find(`.${mainCssClass}`).hasClass(className)).toBe(true);
  });

  it("renders logs", () => {
    const testLogs = [
      {
        id: "1",
        zone: { top: 100, left: 100, width: 100, height: 100 },
      },
      {
        id: "2",
        zone: { top: 200, left: 200, width: 200, height: 200 },
      },
    ];
    const wrapper = mount(
      <Logger logs={testLogs}>
        <div style={{ width: "500px", height: "500px" }}>test div</div>
      </Logger>
    );

    const canvas = wrapper.find("canvas").instance();
    const context = canvas.getContext("2d");

    const path = context.__getPath();
    expect(path).toMatchSnapshot();
  });
});
