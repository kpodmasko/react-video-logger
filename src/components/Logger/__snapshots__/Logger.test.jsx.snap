// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Logger renders logs 1`] = `
Array [
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "beginPath",
  },
  Object {
    "props": Object {
      "height": 200,
      "width": 200,
      "x": 200,
      "y": 200,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "rect",
  },
]
`;
